### PR TITLE
Add support for buildah run with chroot isolation

### PR DESCRIPTION
**Pull request by dfr**
#### What type of PR is this?

> /kind feature

#### What this PR does / why we need it:

This adds support for running containers with chroot isolation. Internally, I actually use a jail instead of a real chroot since this is the only equivalent of the Linux UTS namespace. Large parts of the code are shared with Linux and I moved the shared functions to run_common.go in a similar way to the main run code.

[NO NEW TESTS NEEDED]

#### How to verify it

On a FreeBSD system, run:
```
c=$(sudo buildah from quay.io/dougrabson/freebsd-minimal:13.1)
sudo buildah run --isolation=chroot $c freebsd-version
sudo buildah rm $c
```

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by flouthoc** [comment]
@dfr `cross-build` is breaking because of https://github.com/containers/buildah/pull/4175, maybe you wanna wait and rebase after https://github.com/containers/buildah/pull/4175

**Comment by dfr** [comment]
> @dfr `cross-build` is breaking because of #4175, maybe you wanna wait and rebase after #4175

Makes sense, thanks for the pointer. There were a couple of other cross build breaks related to "freebsd && !cgo" which I fixed.

**Comment by dfr** [comment]
The lint fail is due to my attempt to work around the fact that FreeBSD's struct rlimit contains signed members. I tried to hide this with a type alias but now lint complains on Linux about casting uint64 to uint64. Not sure how to address that yet :(.

**Comment by dfr** [comment]
Rebased

**Comment by rhatdan** [comment]
@flouthoc @nalind @giuseppe @vrothberg @umohnani8 @ashley-cui PTAL

**Comment by rhatdan** [comment]
LGTM

File: chroot/run_freebsd.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	// Create an empty directory for to use for masking directories.
+	/*
+		roEmptyDir := filepath.Join(bundlePath, "empty")
+		if len(spec.Linux.MaskedPaths) > 0 {
+			if err := os.Mkdir(roEmptyDir, 0700); err != nil {
+				return undoBinds, errors.Wrapf(err, "error creating empty directory %q", roEmptyDir)
+			}
+			removes = append(removes, roEmptyDir)
+		}
+	*/
```
Comment (flouthoc) [inline chroot/run_freebsd.go:271]: This should be removed I think.

File: chroot/run_freebsd.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	jconf.Set("persist", false)
+	jconf.Set("path", path)
+	jconf.Set("ip4", jail.INHERIT)
+	jconf.Set("ip6", jail.INHERIT)
+	jconf.Set("allow.raw_sockets", true)
+	jconf.Set("enforce_statfs", 1)
+	_, err := jail.CreateAndAttach(jconf)
+	if err != nil {
+		logrus.Errorf("error creating jail: %v", err)
+		return err
```
Comment (flouthoc) [inline chroot/run_freebsd.go:86]:
```suggestion
		return Errorf("error creating jail: %v", err)
```

File: chroot/run_freebsd.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	jconf.Set("host.hostname", options.Spec.Hostname)
+	jconf.Set("persist", false)
+	jconf.Set("path", path)
+	jconf.Set("ip4", jail.INHERIT)
+	jconf.Set("ip6", jail.INHERIT)
+	jconf.Set("allow.raw_sockets", true)
+	jconf.Set("enforce_statfs", 1)
+	_, err := jail.CreateAndAttach(jconf)
+	if err != nil {
+		logrus.Errorf("error creating jail: %v", err)
```
Comment (flouthoc) [inline chroot/run_freebsd.go:85]:
```suggestion
```

Comment (dfr) [reply chroot/run_freebsd.go]: Thanks!

File: chroot/pty_posix.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+func openpt() (int, error) {
+	fd, err := C.posix_openpt(C.O_RDWR)
+	if err != nil {
+		return -1, err
+	}
+	_, err = C.grantpt(fd)
```
Comment (rhatdan) [inline chroot/pty_posix.go:20]: Please do this on one line
if _, err := c.grantpt(fd); err != nil

File: chroot/pty_posix.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+func ptsname(fd int) (string, error) {
+	path, err := C.ptsname(C.int(fd))
+	if err != nil {
+		return "", err
+	}
+	return C.GoString(path), nil
+}
+
+func unlockpt(fd int) error {
+	_, err := C.unlockpt(C.int(fd))
```
Comment (rhatdan) [inline chroot/pty_posix.go:36]: Please do this on one line
if _, err := c.unlockpt; err != nil

File: chroot/pty_ptmx.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	"golang.org/x/sys/unix"
+)
+
+// Open a PTY using the /dev/ptmx device. The main advantage of using
+// this instead of posix_openpt is that it avoids cgo.
+func getPtyDescriptors() (int, int, error) {
+	// Create a pseudo-terminal -- open a copy of the master side.
+	controlFd, err := unix.Open("/dev/ptmx", os.O_RDWR, 0600)
+	if err != nil {
+		logrus.Errorf("error opening PTY master using /dev/ptmx: %v", err)
```
Comment (rhatdan) [inline chroot/pty_ptmx.go:22]: Do not logrus. Errorf here, Just wrap the err below.

return -1, -1, fmt.Errorf("error opening PTY master using /dev/ptmx: %w", err)

File: chroot/pty_ptmx.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		return -1, -1, err
+	}
+	// Set the kernel's lock to "unlocked".
+	locked := 0
+	if result, _, err := unix.Syscall(unix.SYS_IOCTL, uintptr(controlFd), unix.TIOCSPTLCK, uintptr(unsafe.Pointer(&locked))); int(result) == -1 {
+		logrus.Errorf("error unlocking PTY descriptor: %v", err)
```
Comment (rhatdan) [inline chroot/pty_ptmx.go:28]: No logrus.Errorf, just wrap as above.

File: chroot/pty_ptmx.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		return -1, -1, err
+	}
+	// Get a handle for the other end.
+	ptyFd, _, err := unix.Syscall(unix.SYS_IOCTL, uintptr(controlFd), unix.TIOCGPTPEER, unix.O_RDWR|unix.O_NOCTTY)
+	if int(ptyFd) == -1 {
+		if errno, isErrno := err.(syscall.Errno); !isErrno || (errno != syscall.EINVAL && errno != syscall.ENOTTY) {
+			logrus.Errorf("error getting PTY descriptor: %v", err)
```
Comment (rhatdan) [inline chroot/pty_ptmx.go:35]: No logrus.Errorf, just wrap as above.

File: chroot/pty_ptmx.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+			return -1, -1, err
+		}
+		// EINVAL means the kernel's too old to understand TIOCGPTPEER.  Try TIOCGPTN.
+		ptyN, err := unix.IoctlGetInt(controlFd, unix.TIOCGPTN)
+		if err != nil {
+			logrus.Errorf("error getting PTY number: %v", err)
```
Comment (rhatdan) [inline chroot/pty_ptmx.go:41]: No logrus.Errorf, just wrap as above.

File: chroot/pty_ptmx.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+			return -1, -1, err
+		}
+		ptyName := fmt.Sprintf("/dev/pts/%d", ptyN)
+		fd, err := unix.Open(ptyName, unix.O_RDWR|unix.O_NOCTTY, 0620)
+		if err != nil {
+			logrus.Errorf("error opening PTY %q: %v", ptyName, err)
```
Comment (rhatdan) [inline chroot/pty_ptmx.go:47]: No logrus.Errorf, just wrap as above.

File: chroot/run_freebsd.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+	jconf.Set("host.hostname", options.Spec.Hostname)
+	jconf.Set("persist", false)
+	jconf.Set("path", path)
+	jconf.Set("ip4", jail.INHERIT)
+	jconf.Set("ip6", jail.INHERIT)
+	jconf.Set("allow.raw_sockets", true)
+	jconf.Set("enforce_statfs", 1)
+	_, err := jail.CreateAndAttach(jconf)
+	if err != nil {
+		return fmt.Errorf("error creating jail: %v", err)
```
Comment (rhatdan) [inline chroot/run_freebsd.go:85]: %v->%w

**Review by rhatdan — APPROVED** [review]

**Comment by rhatdan** [comment]
LGTM
@flouthoc @nalind @vrothberg @giuseppe PTAL

**Comment by dfr** [comment]
Rebased

**Review by flouthoc — APPROVED** [review]
LGTM @nalind PTAL

**Comment by dfr** [comment]
Rebased

**Comment by dfr** [comment]
Rebased. I'm hoping this can be merged sometime soonish. I know its not particularly important given the existing support for isolation=oci but a very early version of this code was used very early on to run simple container tests before I had an OCI runtime. This helped convince me that porting the containers stack to FreeBSD was viable and I probably wouldn't be hassling you all with podman PRs without it so it has some sentimental value :).

**Comment by dfr** [comment]
Rebased

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Review by flouthoc — APPROVED** [review]
LGTM
/lgtm
/approve

**Pull request merged**